src/WebJobs.Script.WebHost/Diagnostics/DeferredLogger.cs

```diff
@@ -0,0 +1,50 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.Azure.WebJobs.Script.Eventing;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost.Diagnostics
+{
+    internal class DeferredLogger : ILogger
+    {
+        private readonly string _category;
+        private readonly IScriptEventManager _eventManager;
+
+        public DeferredLogger(string category, IScriptEventManager eventManager)
+        {
+            _category = category;
+            _eventManager = eventManager;
+        }
+
+        public IDisposable BeginScope<TState>(TState state) => DictionaryLoggerScope.Push(state);
+
+        public bool IsEnabled(LogLevel logLevel)
+        {
+            return true;
+        }
+
+        public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception exception, Func<TState, Exception, string> formatter)
+        {
+            string StateFormatter(object s, Exception e)
+            {
+                TState castState = (TState)s;
+                return formatter?.Invoke(castState, e);
+            }
+
+            DeferredLogMessage logMessage = new DeferredLogMessage
+            {
+                Timestamp = DateTimeOffset.UtcNow,
+                Category = _category,
+                LogLevel = logLevel,
+                EventId = eventId,
+                State = state,
+                Exception = exception,
+                Formatter = StateFormatter
+            };
+
+            _eventManager.Publish(new DeferredLogEvent(logMessage));
```

Comment:
So this deferred logger is registered early and passes logs to event manager. Where does the deferred buffering happen? Where is the flush? When does this deferred logger get removed once not needed anymore?

Reply:
I've swapped this out to use a `BufferBlock`. I was hoping I could get the buffering/flushing to work with Reactive but wasn't able to. Going to do some private site extension testing and then will push up the updated PR.